### PR TITLE
fix(rust/gui-client): when the Client starts with a token but no Internet, wait for Internet and then connect

### LINKED ISSUE #6389: bug(rust/gui-client): Client will never sign in if started with no Internet
```[tasklist]
### Tasks
- [x] Open PR to make the error louder (#6409)
- [x] Check if macOS has this problem (It does)
- [x] Check if Android has this problem
- [x] Check if iOS has this problem
- [x] Open PR to retry the connection whenever the network / DNS changes (#6414)
```

#### Describe the bug

If the GUI Client starts without Internet, it will never sign in and never raise the tunnel. This is because `PhoenixChannel::connect` fails and the Client never retries it

- Headless Client - This is intended behavior, PR #6409 documents it
- Rust GUI Client - This is a UX bug, PR #6409 changes it to a user-facing error
- macOS Client - If you start with a token and no Internet, the macOS Client appears signed out. When you gain Internet, you are still signed out. When you then restart the Client, you sign in automatically
- Android - Restart the phone in Airplane mode to close and re-open Firezone. Then disable Airplane mode. Firezone will not work until you open the app.
- iOS - Restart the phone in Airplane to close and re-open Firezone. Then disable Airplane mode. Firezone will not work until you enable the VPN config.

#### To Reproduce

1. Sign in to the Client
2. Quit the Client
3. Disconnect from the Internet
4. Start the Client
5. Wait 10 seconds or so
6. Connect to the Internet

The IPC service logs will have a bunch of errors in them and the Client will never sign in.

The headless Client has a similar problem, it bails out if you start it with no Internet.

#### Expected behavior

- As a user, expected the sign-in to resume once Internet is gained
- As the Client app, expected the IPC service to respond with `on_disconnect` or something if requests are failing because there's no connlib session
- As the IPC service, expected an `on_disconnect` from connlib if the portal connection fails permanently

Otherwise this will interfere with auto-starting the GUI, since users will have to reset the connection if the computer started without Internet.

#### Screenshots / Logs

Bad log:

```
2024-08-21T17:05:34.294353Z  INFO firezone_headless_client::ipc_service::ipc::platform: Accepted IPC connection client_pid=12324
2024-08-21T17:05:34.295758Z DEBUG firezone_headless_client::device_id: Loaded device ID from disk id="0a9ba5c0-7edb-4f3a-a2a1-b4d011755849"
2024-08-21T17:05:34.296576Z ERROR firezone_headless_client::ipc_service: Error while handling IPC message from client error=No such host is known. (os error 11001)
2024-08-21T17:05:34.304275Z ERROR firezone_headless_client::ipc_service: Error while handling IPC message from client error=No connlib session
2024-08-21T17:05:47.953716Z ERROR firezone_headless_client::ipc_service: Error while handling IPC message from client error=No connlib session
2024-08-21T17:05:48.454505Z ERROR firezone_headless_client::ipc_service: Error while handling IPC message from client error=No connlib session
```

The last few errors are from the GUI trying to tell the IPC service about network / DNS change events, but the IPC service won't listen because it's not signed in.

Good: (If the GUI started with Internet)

```
2024-08-21T17:07:20.441786Z  INFO firezone_headless_client::ipc_service::ipc::platform: Accepted IPC connection client_pid=15376
2024-08-21T17:07:20.443573Z DEBUG firezone_headless_client::device_id: Loaded device ID from disk id="0a9ba5c0-7edb-4f3a-a2a1-b4d011755849"
2024-08-21T17:07:20.486350Z DEBUG phoenix_channel: Connecting to portal host=api.firez.one user_agent=Windows/10.0.22631 connlib/1.2.0 (x86_64)
2024-08-21T17:07:20.486627Z DEBUG new{mtu=1280}: firezone_bin_shared::tun_device_manager::windows: wintun.dll path path="C:\\Windows\\system32\\config\\systemprofile\\AppData\\Local\\dev.firezone.client\\data\\wintun.dll"
2024-08-21T17:07:20.493580Z DEBUG firezone_bin_shared::windows: Resolved best route outside of tunnel interface src=192.168.1.12 dst=34.160.64.79
2024-08-21T17:07:20.623380Z  INFO phoenix_channel: Connected to portal host=api.firez.one
2024-08-21T17:07:20.686105Z DEBUG firezone_tunnel::client: Updating DNS servers mapping={100.100.111.1 <> 1.1.1.1:53, fd00:2021:1111:8000:100:100:111:0 <> [2606:4700:4700::1111]:53}
```

#### Platform (please complete the following information)

- _Component (i.e. macOS client / Linux client / Gateway / Admin portal):_ GUI Client, macOS Client, possibly others if this is a connlib issue
- _Firezone Version (e.g. 1.0.0 or N/A):_ 16da501 (tip of `main`)
- _OS and version: (e.g. Ubuntu 22.04 or N/A):_ Windows 11, probably Ubuntu, possibly others
- _Deployment method: (e.g. Docker / Systemd / App Store or N/A):_ CI-built MSI

#### Additional context

Discovered while testing #6382

### PR DESCRIPTION
**Pull request by ReactorScram**
Closes #6389 

I added a retry button since the network change detection is flaky inside Parallels. On bare metal Windows it works fine.

### CODE REVIEW

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 1]
```diff
@@ -734,17 +755,68 @@ impl Controller {
...
-                self.tray.set_icon(system_tray::Icon::SignedOut).ok();
-                Err(Error::IpcServiceTerminating)
+            Err(IpcServiceError::PortalConnection(error)) => {
+                tracing::warn!(
+                    ?error,
+                    "Failed to connect to Firezone Portal, will try again when the network changes"
+                );
+                if first_time {
+                    os::show_notification(
+                        "Firezone offline",
```
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/gui.rs]:
```suggestion
                        "Internet offline",
```

I wonder if people will think this means the Firezone service is offline

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 2]
```diff
@@ -901,17 +969,31 @@ async fn run_controller(
...
-                    let resolvers = firezone_headless_client::dns_control::system_resolvers_for_gui()?;
+                match &controller.status {
+                    Status::Disconnected => {}
+                    Status::TunnelReady { .. } | Status::WaitingForPortal { .. } | Status::WaitingForTunnel { .. } => {
+                        let resolvers = firezone_headless_client::dns_control::system_resolvers_for_gui()?;
                     tracing::debug!(?resolvers, "New DNS resolvers, calling `Session::set_dns`");
                     controller.ipc_client.set_dns(resolvers).await?;
+                    },
+                    Status::WaitingForNetwork { token } => {
+                        tracing::debug!("New DNS resolvers, retrying Portal connection...");
```
Comment (jamilbk) [inline rust/gui-client/src-tauri/src/client/gui.rs]: Does this effectively mean we trigger on DNS changes? I suppose on Windows that's pretty reliable, but what's the behavior on Linux?

Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs]: It should be DNS changes and network changes on both OSes
- Linux DNS via D-Bus https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/linux.rs#L40
- Linux network via D-Bus https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/linux.rs#L61
- Windows DNS via registry listener https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/windows.rs#L381
- Windows network via complicated COM callback https://github.com/firezone/firezone/blob/4ec78f70c220f564e143ed4667543cd58a2baa6b/rust/bin-shared/src/network_changes/windows.rs#L122thing

Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs]: The only alternative I could think of was polling which didn't sound pleasant

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 3]
```diff
@@ -452,10 +452,27 @@ pub(crate) enum ControllerRequest {
...
+        /// The instant when we sent our most recent connect request.
+        start_instant: Instant,
+        /// The token to log in to the Portal, in case we need to retry the connection request.
+        token: SecretString,
+    },
+    /// Firezone has connected to the Portal and is raising the tunnel.
+    WaitingForTunnel {
+        /// The instant when we sent our most recent connect request.
+        start_instant: Instant,
+    },
```
Comment (thomaseizinger) [inline rust/gui-client/src-tauri/src/client/gui.rs]: Can we perhaps merge some of these together and make the logic on what the GUI does a bit more general?

1. On start-up, try to sign in. Show to user as "Connecting".
2. If sign-in fails due to any reason that is not an authentication error, still show "Connecting" but internally move to a `ConnectFailed` state that holds the token.
3. On network changes, try to send another connect when in that state.
4. Optionally: Provide a way for the user to force another connect by maybe adding a "Retry now" option to the "Connecting" element in the GUI.

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 4]
```diff
@@ -734,17 +755,68 @@ impl Controller {
...
+            Err(IpcServiceError::PortalConnection(error)) => {
+                tracing::warn!(
+                    ?error,
+                    "Failed to connect to Firezone Portal, will try again when the network changes"
+                );
+                if first_time {
+                    os::show_notification(
+                        "Internet offline",
+                        "Firezone will connect automatically when you have Internet.",
+                    )?;
```
Comment (thomaseizinger) [inline rust/gui-client/src-tauri/src/client/gui.rs]: Is that actually useful? As a user, this doesn't really tell me anything that I don't already know. We are already doing the right thing by attempting to reconnect in the background. Grabbing the users attention here feels noisy.

**Review by thomaseizinger — APPROVED** [review]
I left some ideas on how we can potentially simplify the state machine. Concept ACK though

File: rust/gui-client/src-tauri/src/client/gui.rs (hunk 4 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs]: You're right, if they don't have Internet they won't care that they can't access Resources

File: rust/gui-client/src-tauri/src/client/gui.rs (hunk 3 above)
Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs]: So generalize `WaitingForNetwork` to `RetryingConnection`? 

I prefer naming states for either what their next step is ("Waiting for X") or at least what they're doing ("Retrying"), if it's `ConnectFailed` it doesn't say where it's going

Comment (thomaseizinger) [reply rust/gui-client/src-tauri/src/client/gui.rs]: > I prefer naming states for either what their next step is ("Waiting for X") or at least what they're doing ("Retrying"), if it's `ConnectFailed` it doesn't say where it's going

More complex state machines will potentially:

- Transition to different states
- Offer multiple things that can be done

What they still have in common is that multiple transitions lead to the same state. In that case, naming it after what happened can be a good middle-ground (`SignedIn`, `ConnectFailed`).

I don't mind either way, `RetryingConnection` also works for me :)

File: rust/gui-client/src-tauri/src/client/gui.rs [hunk 5]
```diff
@@ -452,10 +452,25 @@ pub(crate) enum ControllerRequest {
...
-    Connecting { start_instant: Instant },
+    /// At least one connection request has failed, due to failing to reach the Portal, and we are waiting for a network change before we try again
+    RetryingConnection {
+        /// The token to log in to the Portal, for retrying the connection request.
+        token: SecretString,
+    },
     /// Firezone is ready to use.
     TunnelReady { resources: Vec<ResourceDescription> },
+    /// Firezone is signing in to the Portal.
+    WaitingForPortal {
```
Comment (thomaseizinger) [inline rust/gui-client/src-tauri/src/client/gui.rs:463]: What do we gain from splitting `Connecting` into the two `Waiting` states? `connlib` will internally reconnect to the portal if it fails whilst we have the tunnel.

Currently, the fact that we resolve the IPs on startup is somewhat of an implementation detail and that may change again. For example, if we end up doing something like https://github.com/firezone/firezone/pull/6427, we may resolve them dynamically on startup and thus, creating phoenix-channel will become infallible again.

I would suggest to not let this implementation detail escape the lower layers and just treat the entire time as `Connecting`.

Comment (ReactorScram) [reply rust/gui-client/src-tauri/src/client/gui.rs:463]: It doesn't affect the behavior, but it's nice to see "Connecting to Portal" in the menu if we're trying to debug something

Comment (thomaseizinger) [reply rust/gui-client/src-tauri/src/client/gui.rs:463]: If it is just for debugging, could we also achieve the same with logs? My recommendation would be to not treat the initial boot-up of connlib as two phases. Where it failed in its boot-process shouldn't really matter to the clients.

The above state machine kind of suggests that we would be also showing this when we are internally re-connecting to the portal but we don't. That is completely transparent to the client (until we do https://github.com/firezone/firezone/issues/5557).

**Pull request merged**